### PR TITLE
Allow BatchNormalization to be used as a shared layer

### PR DESCRIPTION
**Pull request by henry0312**
Before adding `mode=2` ( fchollet/keras@88d523e01b68a87298d14dac1d2cee864bf5a22a and fchollet/keras@601f3e7cdb750a3cc9481c548327d4e09da4ea5a ), BN updated its parameters with last input batch when using BN as shared layer.
I think the two commits are right (unable to use BN as a shared layre), but we can use BN as a shared layer for some cases.

For example, when two inputs are the same, BN will calculate correct mean and std.
Additonaly, if there are some data flows but the data sorce is the same (e.g. all batches created from MNIST), BN will be able to calculate correct those after enough training.

Therefore, I added `allow_shared` option to BN, which allows this layer to be used as shared layer if True when `mode=0`.

### CODE REVIEW

File: keras/layers/normalization.py [hunk 1]
```diff
@@ -26,6 +26,8 @@ class BatchNormalization(Layer):
             - 2: feature-wise normalization, like mode 0, but
                 using per-batch statistics to normalize the data during both
                 testing and training.
+        allow_shared: bool.
+            if True, allow this layer to be used as a shared layer when mode = 0.
```
Comment (fchollet) [inline keras/layers/normalization.py]: The weight update behavior should be explained.

File: keras/layers/normalization.py [hunk 2]
```diff
@@ -55,12 +57,13 @@ class BatchNormalization(Layer):
     # References
         - [Batch Normalization: Accelerating Deep Network Training by Reducing Internal Covariate Shift](http://jmlr.org/proceedings/papers/v37/ioffe15.html)
     '''
-    def __init__(self, epsilon=1e-6, mode=0, axis=-1, momentum=0.9,
+    def __init__(self, epsilon=1e-6, mode=0, allow_shared=False, axis=-1, momentum=0.9,
```
Comment (fchollet) [inline keras/layers/normalization.py]: I would name it "allow_sharing".

Comment (fchollet) [inline keras/layers/normalization.py]: PS it should go at the end of the arguments list for backwards compatibility.

**Comment by henry0312** [comment]
Updated.

**Comment by henry0312** [comment]
Updated :D

**Comment by fchollet** [comment]
One problem with merging this PR is that BN sharing simply *does not* work with nested models or with TimeDistributed. Not just that it doesn't have a sensible behavior --it actually crashes. Merging this PR would reintroduce this bug.

e.g. https://github.com/fchollet/keras/issues/2827

**Comment by henry0312** [comment]
I see...🤔

I want to decrease crashes as much as possible.
Therefore, we can take the below two measures:

First, we doesn't make the exception message include `allow_sharing`.
i.e.  the below changes will not be included:

```diff
                     raise Exception('You are attempting to share a '
                                     'same `BatchNormalization` layer across '
                                     'different data flows. '
                                     'This is not possible. '
-                                    'You should use `mode=2` in '
+                                    'You should use `mode=2` or `allow_sharing=True` in '
                                     '`BatchNormalization`, which has '
                                     'a similar behavior but is shareable '
                                     '(see docs for a description of '
```

Then, we rename `allow_sharing` to something like `advanced_sharing`, which means that the only users that understand the behavior can use it.

What do you think?

**Comment by henry0312** [comment]
The point is that we want to prevent crashes, on the other hand, this PR is not wrong in some cases...

**Comment by fchollet** [comment]
I get it, but it remains that it is formally incorrect. The exception as it is implemented is meant to prevent formally incorrect code to be executed, not just meant to prevent unexpected runtime behaviors. Overriding the exception opens the door to both.

**Comment by henry0312** [comment]
> The exception as it is implemented is meant to prevent formally incorrect code to be executed, not just meant to prevent unexpected runtime behaviors. 

I think so.
What am I supposed to do?

**Comment by henry0312** [comment]
I made a change.

```diff
diff --git a/keras/layers/normalization.py b/keras/layers/normalization.py
index 81207ad..d4904b6 100644
--- a/keras/layers/normalization.py
+++ b/keras/layers/normalization.py
@@ -43,9 +43,11 @@ class BatchNormalization(Layer):
             [initializations](../initializations.md)), or alternatively,
             Theano/TensorFlow function to use for weights initialization.
             This parameter is only relevant if you don't pass a `weights` argument.
-        allow_sharing: bool.
+        advanced_sharing: bool.
             if True, allow this layer to be used as a shared layer in mode 0.
             The weights will be updated with the last input batch.
+            Note that BatchNormalization sharing simply does not work with nested models
+            or with TimeDistributed.
 
     # Input shape
         Arbitrary. Use the keyword argument `input_shape`
@@ -59,12 +61,12 @@ class BatchNormalization(Layer):
         - [Batch Normalization: Accelerating Deep Network Training by Reducing Internal Covariate Shift](http://jmlr.org/proceedings/papers/v37/ioffe15.html)
     '''
     def __init__(self, epsilon=1e-6, mode=0, axis=-1, momentum=0.9,
-                 weights=None, beta_init='zero', gamma_init='one', allow_sharing=False, **kwargs):
+                 weights=None, beta_init='zero', gamma_init='one', advanced_sharing=False, **kwargs):
         self.beta_init = initializations.get(beta_init)
         self.gamma_init = initializations.get(gamma_init)
         self.epsilon = epsilon
         self.mode = mode
-        self.allow_sharing = allow_sharing
+        self.advanced_sharing = advanced_sharing
         self.axis = axis
         self.momentum = momentum
         self.initial_weights = weights
@@ -116,12 +118,14 @@ class BatchNormalization(Layer):
                 out = K.reshape(self.gamma, broadcast_shape) * x_normed + K.reshape(self.beta, broadcast_shape)
             else:
                 # mode 0
-                if self.called_with not in {None, x} and self.allow_sharing is False:
+                if self.called_with not in {None, x} and self.advanced_sharing is False:
                     raise Exception('You are attempting to share a '
                                     'same `BatchNormalization` layer across '
                                     'different data flows. '
                                     'This is not possible. '
-                                    'You should use `mode=2` or `allow_sharing=True` in '
+                                    'You should use `mode=2` (if it is used with '
+                                    'nested models or TimeDistributed) or '
+                                    '`advanced_sharing=True` in '
                                     '`BatchNormalization`, which has '
                                     'a similar behavior but is shareable '
                                     '(see docs for a description of '
@@ -153,6 +157,6 @@ class BatchNormalization(Layer):
                   "mode": self.mode,
                   "axis": self.axis,
                   "momentum": self.momentum,
-                  "allow_sharing": self.allow_sharing}
+                  "advanced_sharing": self.advanced_sharing}
         base_config = super(BatchNormalization, self).get_config()
         return dict(list(base_config.items()) + list(config.items()))
```

There are 2 points.

1. the default value of `advanced_sharing` is **False**  
    this prevents users from facing unnecessary crashes.
2. explaining that what `advancecd_sharing` isn't able to do  
    this allows users to know which to use `mode=2` or `advanced_shring`.

I guess this change will solve almost all of problems in the PR.
How do you feel about this?

**Comment by henry0312** [comment]
updated (applied the above).

**Comment by fchollet** [comment]
All things considered: let's not do it. You can maintain this change in your own version of Keras, for your own usage. I don't think it is widely useful.

**Comment by henry0312** [comment]
All right. Thank you for your review.

> I don't think it is widely useful.

It's true that almost all of users are satisfied with the current BatchNormalization and it's a bit hard to understand using one as a shared layer.

**Comment by konstantinberlin** [comment]
I am trying to understand the current BatchNormalization code after the introduction of mode=2. Is it correct that during the prediction phase the prediction is made using the aggregate statistic of the given batch (so mean and std is computed from prediction batch). So if, lets say, I make a prediction using a large batch size, it would be one value, but if I make a forward prediction on a single value, it could swing wildly to a different value?

Aka prediction is stochastic, rather than deterministic.
